src/workerd/io/worker.c++

```diff
@@ -2317,8 +2301,243 @@ public:
+      // (e.g. resume).
+      return incomingQueue.when(
+        [](const MessageQueue& incomingQueue) {
+          return (incomingQueue.head < incomingQueue.messages.size() ||
+                  incomingQueue.status == MessageQueue::Status::CLOSED);
+        },
+        [](MessageQueue& incomingQueue) -> kj::Maybe<kj::String> {
+          if (incomingQueue.status == MessageQueue::Status::CLOSED) return {};
+          return pollMessage(incomingQueue);
+        });
+    }
+
+    kj::Promise<void> messagePump() {
+      // Message pumping promise that should be evaluated on the InspectorService
+      // thread.
+      return incomingLoop().exclusiveJoin(outgoingLoop());
+    }
+
+    void send(kj::String message) {
+      auto lockedOutgoingQueue = outgoingQueue.lockExclusive();
+      if (lockedOutgoingQueue->status == MessageQueue::Status::CLOSED) return;
+      lockedOutgoingQueue->messages.add(kj::mv(message));
+      outgoingQueueNotifier->notify();
+    }
+
+  private:
+    struct MessageQueue {
+      kj::Vector<kj::String> messages;
+      size_t head;
+      enum class Status { ACTIVE, PAUSED, CLOSED } status;
+    };
+
+    static kj::Maybe<kj::String> pollMessage(MessageQueue& messageQueue) {
+      if (messageQueue.head < messageQueue.messages.size()) {
+        kj::String message = kj::mv(messageQueue.messages[messageQueue.head++]);
+        if (messageQueue.head == messageQueue.messages.size()) {
+          messageQueue.head = 0;
+          messageQueue.messages.clear();
+        }
+        return kj::mv(message);
+      }
+      return {};
+    }
+
+    static void deliverProtocolMessages(kj::Maybe<InspectorChannelImpl&>& channel,
+                                        kj::Locked<MessageQueue>& incomingQueue) {
+      KJ_REQUIRE(incomingQueue->status == MessageQueue::Status::ACTIVE);
+      KJ_IF_MAYBE(c, channel) {
+        kj::ArrayPtr<kj::String> messages = incomingQueue->messages.slice(
+            incomingQueue->head, incomingQueue->messages.size());
+        c->dispatchProtocolMessages(messages);
+        incomingQueue->messages.clear();
+        incomingQueue->head = 0;
+      }
+    }
+
+    void shutdown() {
+      // Wake any waiters.
+      outgoingQueueNotifier->notify();
+
+      // Drain incoming queue, the isolate thread may be waiting on it
+      // on will notice it is closed if woken without any messages to
+      // deliver in WebSocketIoWorker::waitForMessage().
+      {
+        auto lockedIncomingQueue = incomingQueue.lockExclusive();
+        lockedIncomingQueue->head = 0;
+        lockedIncomingQueue->messages.clear();
+        lockedIncomingQueue->status = MessageQueue::Status::CLOSED;
+      }
+      {
+        auto lockedOutgoingQueue = outgoingQueue.lockExclusive();
+        lockedOutgoingQueue->status = MessageQueue::Status::CLOSED;
+      }
+    }
+
+    kj::Promise<void> incomingLoop() {
+      return webSocket.receive().then([this](kj::WebSocket::Message&& message) -> kj::Promise<void> {
```

Comment:
Optional: Might be nice to convert to a coroutine with a for() loop? Also for `outgoingLoop()`.

Reply:
Done, but PTAL as it's a first conversion. Thanks.